R/credentials_gce.R

```diff
@@ -114,36 +138,112 @@ credentials_gce <- function(scopes = "https://www.googleapis.com/auth/cloud-plat
   },
-  #' @description Placeholder implementation of required method
+  #' @description Request an access token.
   init_credentials = function() {
-    self$credentials <- list(access_token = NULL)
+    gargle_debug("GceToken init_credentials")
+    token <- fetch_gce_access_token(
+      self$params$scope,
+      service_account = self$params$service_account
+    )
+
+    # find out the scopes actually obtained
+    # https://www.googleapis.com/oauth2/v3/tokeninfo
+    req <- request_build(
+      method = "GET",
+      path = "oauth2/v3/tokeninfo",
+      params = list(access_token = token$access_token),
+      base_url = "https://www.googleapis.com"
+    )
+    resp <- request_make(req)
+    info <- response_process(resp)
+    actual_scopes <- normalize_scopes(strsplit(info$scope, split = "\\s+")[[1]])
+
+    missing <- setdiff(self$params$scope, actual_scopes)
+    if (length(missing) > 0) {
+      gargle_debug(c(
+        "!" = "{cli::qty(length(missing))}{?This/These} requested \\
+             scope{?s} {?is/are} not among the scopes for the \\
+             access token returned by the metadata server:",
+        bulletize(missing, bullet = "x"),
+        "i" = "If there are problems downstream, this might be the root cause."
+      ))
+      gargle_debug(c(
+        "!" = "Setting token scopes to:",
+        bulletize(actual_scopes)
+      ))
+      self$params$scope <-actual_scopes
+    }
+
+    self$credentials <- token
+    self
   },
-  #' @description Placeholder implementation of required method
-  cache = function(...) {},
-  #' @description Placeholder implementation of required method
-  load_from_cache = function(...) {},
-  #' @description Placeholder implementation of required method
+  #' @description Refreshes the token. In this case, that just means "ask again
+  #'   for an access token".
+  refresh = function() {
+    gargle_debug("GceToken refresh")
+    # There's something kind of wrong about this, because it's not a true
+    # refresh. But this method is basically required by the way httr currently
+    # works.
+    # This means that some uses of $refresh() aren't really appropriate for a
+    # GceToken.
+    # For example, if I attempt token_userinfo(x) on a GceToken that lacks
+    # appropriate scope, it fails with 401.
+    # httr tries to "fix" things by refreshing the token. But this is
+    # not a problem that refreshing can fix.
+    # I've now prevented an explicit refresh in token_userinfo(), but an
+    # implicit one still eventually happens in httr:::request_perform().
+    self$init_credentials()
+  },
+  #' @description Placeholder implementation of required method. Returns `TRUE`.
   can_refresh = function() {
     TRUE
   },
-  #' @description Refresh a GCE token
-  refresh = function() {
-    # The access_token can only include the token itself, not the expiration and
-    # type. Otherwise, the httr code will create extra header lines that bust
-    # the POST request:
```

Comment:
A bit of fussiness was introduced in #98 that I now think was a red herring. I merged it, but at the time I had very little understanding of this flow and definitely had no hands-on experience with it (whereas I do now).

I'm getting rid of this fussiness now.